### ISSUE TITLE
r/aws_batch_job_definition must always be replaced in 2.48

### ISSUE THREAD

**Issue by theophilechevalier** [terraform-providers/terraform-provider-aws#11998]
Hi, I got stuck in a bug using the provider in version `2.48`. `aws_batch_job_definition` is always recreated.

Introduced in https://github.com/terraform-providers/terraform-provider-aws/pull/11488

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version
```
Terraform v0.12.20
+ provider.aws v2.48.0
```

### Affected Resource(s)

* aws_batch_job_definition

### Terraform Configuration Files

```hcl
terraform {
  backend "local" {
    path = ".terraform/state/terraform.tfstate"
  }
}

provider "aws" {
  region = "eu-central-1"

  version = "~> 2.40"
}

resource "aws_iam_role" "role" {
  name = "test_example_role"

  assume_role_policy = <<EOF
{
    "Version": "2012-10-17",
    "Statement": [
    {
        "Action": "sts:AssumeRole",
        "Effect": "Allow",
        "Principal": {
        "Service": "ecs-tasks.amazonaws.com"
        }
    }
    ]
}
EOF
}

resource "aws_ecr_repository" "ecr" {
  name = "foo_ecr"
}

resource "aws_batch_job_definition" "test" {
  name = "test"
  type = "container"

  container_properties = <<CONTAINER_PROPERTIES
{
    "command": ["start.py", "Ref::S3bucket", "Ref::S3key"],
    "image": "${aws_ecr_repository.ecr.name}:image",
    "memory": 2048,
    "vcpus": 8,
    "jobRoleArn": "${aws_iam_role.role.arn}"
}
CONTAINER_PROPERTIES
}
```

### Debug Output
https://gist.github.com/theophilechevalier/5b0421ef8ca161c0a7ab82fe630b6179

### Expected Behavior

Resource should not be always recreated.

### Actual Behavior
Resource is always recreated.

### Steps to Reproduce

1. `terraform init`
2. `terraform apply`
3. `terraform plan`

### Important Factoids

It works as expected with a provider version `<= 2.47`, so it's a regression.

**Comment by bflad** [comment]
Hi @theophilechevalier 👋 Thank you for submitting this and sorry you ran into trouble. Just to double check, could you please show the `terraform plan` output for that resource?

It looks like the API may canonicalize to including additional arrays as empty arrays, in addition to resource requirements:

```
"volumes":[],"environment":[],"mountPoints":[],"ulimits":[],"resourceRequirements":[]
```

**Comment by theophilechevalier** [comment]
Hi @bflad, thanks for the (very) quick reply!

Here is the output of the plan:
```
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.

aws_ecr_repository.ecr: Refreshing state... [id=foo_ecr]
aws_iam_role.role: Refreshing state... [id=test_example_role]
aws_batch_job_definition.test: Refreshing state... [id=arn:aws:batch:eu-central-1:129757167202:job-definition/test:7]

------------------------------------------------------------------------

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
-/+ destroy and then create replacement

Terraform will perform the following actions:

  # aws_batch_job_definition.test must be replaced
-/+ resource "aws_batch_job_definition" "test" {
      ~ arn                  = "arn:aws:batch:eu-central-1:129757167202:job-definition/test:7" -> (known after apply)
      ~ container_properties = jsonencode(
          ~ {
                command              = [
                    "start.py",
                    "Ref::S3bucket",
                    "Ref::S3key",
                ]
              - environment          = [] -> null
                image                = "foo_ecr:image"
                jobRoleArn           = "arn:aws:iam::129757167202:role/test_example_role"
                memory               = 2048
              - mountPoints          = [] -> null
              - resourceRequirements = [] -> null
              - ulimits              = [] -> null
                vcpus                = 8
              - volumes              = [] -> null
            } # forces replacement
        )
      ~ id                   = "arn:aws:batch:eu-central-1:129757167202:job-definition/test:7" -> (known after apply)
        name                 = "test"
      - parameters           = {} -> null
      ~ revision             = 7 -> (known after apply)
        type                 = "container"
    }

Plan: 1 to add, 0 to change, 1 to destroy.

------------------------------------------------------------------------

Note: You didn't specify an "-out" parameter to save this plan, so Terraform
can't guarantee that exactly these actions will be performed if
"terraform apply" is subsequently run.
```

**Comment by bflad** [comment]
Perfect, @theophilechevalier 👍  Should be able to submit a fix for this fairly quickly.

**Comment by bflad** [comment]
Fix submitted: https://github.com/terraform-providers/terraform-provider-aws/pull/12000

**Comment by bflad** [comment]
The fix for this has been merged and will release with version 2.49.0 of the Terraform AWS Provider, tomorrow. 👍
